### PR TITLE
Add Cloudflare DNS01

### LINKED ISSUE #159: Feature: Offer a DNS01 provider for Cloudflare
Some clients need to use Cloudflare for their DNS configuration with cert-manager.

## Expected Behaviour

A feature like DO / Route53 / Google Cloud DNS, but for Cloudflare

## Current Behaviour

Users can configure it manually or via YAML

## Possible Solution

Examine the code / approach for DO or Route53 and replicate for Cloudflare.

If help is needed with testing please shout when the PR is up.

The configuration looks quite simple:

https://cert-manager.io/docs/configuration/acme/dns01/cloudflare/

### PR DESCRIPTION
**Pull request by alexellis**
## Description

Add Cloudflare DNS01 - Fixes #159 

## How Has This Been Tested?

Tested with unit tests for feature detection in the YAML file and for generation of the YAML spec as per https://cert-manager.io/docs/configuration/acme/dns01/cloudflare/

A Cloudflare user could now test this e2e and report back.

## Checklist:

I have:

- [x] checked my changes follow the style of the existing code / OpenFaaS repos
- [x] updated the documentation and/or roadmap in README.md
- [x] read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] signed-off my commits with `git commit -s`
- [x] added unit tests

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -395,13 +395,14 @@ OpenFaaS Cloud can use cert-manager to automatically provision TLS certificates
 
 > This feature is optional, but highly recommended
 
-Pick between the following providers for the DNS01 challenge:
+Pick between the following providers for the [DNS01 challenge](https://cert-manager.io/docs/configuration/acme/dns01/):
 
 * DigitalOcean DNS (free at time of writing)
```
Comment (Waterdrips) [inline README.md]: Slightly side note, but is this still free?

Comment (alexellis) [reply README.md]: Yes

**Pull request merged**